### PR TITLE
feat: Added Warehouse Type master

### PR DESCRIPTION
**Pull request by Anurag810**
Added master **Warehouse Type** to **define Warehouse Type in Warehouse.**
Which is also used as **Filter** in **Report Stock Balance**

**Warehouse**
Screenshot 2019-05-22 at 11 32 43 AM

**Report Stock Balance**
Screenshot 2019-05-22 at 11 35 24 AM

### CODE REVIEW

File: erpnext/stock/doctype/warehouse_type/warehouse_type.json [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+   "fieldtype": "Small Text",
+   "label": "Description"
+  }
+ ],
+ "modified": "2019-05-21 15:30:06.754938",
+ "modified_by": "Administrator",
+ "module": "Stock",
+ "name": "Warehouse Type",
+ "owner": "Administrator",
+ "permissions": [
```
Comment (nabinhait) [inline erpnext/stock/doctype/warehouse_type/warehouse_type.json]: Give permissions to Stock Manager

File: erpnext/stock/report/stock_balance/stock_balance.js [hunk 2]
```diff
@@ -49,7 +49,22 @@ frappe.query_reports["Stock Balance"] = {
 			"label": __("Warehouse"),
 			"fieldtype": "Link",
 			"width": "80",
-			"options": "Warehouse"
+			"options": "Warehouse",
+			get_query: () => {
```
Comment (nabinhait) [inline erpnext/stock/report/stock_balance/stock_balance.js]: Filter by warehouse type only if type is selected, otherwise show all warehouses.

**Review by nabinhait — CHANGES_REQUESTED** [review]
Filtering warehouse by warehouse type is not the expectation. 

The expectation is, if warehouse type is selected (and warehouse not selected), in server side for calculating stock balance consider warehouses based on warehouse type.

File: erpnext/stock/report/stock_balance/stock_balance.js (hunk 2 above)
Comment (Anurag810) [reply erpnext/stock/report/stock_balance/stock_balance.js]: done

File: erpnext/stock/doctype/warehouse_type/warehouse_type.json (hunk 1 above)
Comment (Anurag810) [reply erpnext/stock/doctype/warehouse_type/warehouse_type.json]: Added permission on the basis of warehouse master

**Comment by Anurag810** [comment]
> Filtering warehouse by warehouse type is not the expectation.
> 
> The expectation is, if warehouse type is selected (and warehouse not selected), in server side for calculating stock balance consider warehouses based on warehouse type.

done

File: erpnext/stock/report/stock_balance/stock_balance.py [hunk 3]
```diff
@@ -257,6 +257,9 @@ def get_item_reorder_details(items):
 	return dict((d.parent + d.warehouse, d) for d in item_reorder_details)
 
 def validate_filters(filters):
+	if filters.warehouse_type and not filters.warehouse:
+		frappe.throw(_("Please set Warehouse"))
```
Comment (nabinhait) [inline erpnext/stock/report/stock_balance/stock_balance.py]: Don't make warehouse mandatory. Instead, consider all warehouses with the selected warehouse type.

**Review by nabinhait — CHANGES_REQUESTED** [review]

File: erpnext/stock/doctype/warehouse_type/warehouse_type.json (hunk 1 above)
Comment (nabinhait) [reply erpnext/stock/doctype/warehouse_type/warehouse_type.json]: Also give write/create permission to Stock Manager role

**Review by nabinhait — CHANGES_REQUESTED** [review]

File: erpnext/stock/report/stock_balance/stock_balance.py (hunk 3 above)
Comment (Anurag810) [reply erpnext/stock/report/stock_balance/stock_balance.py]: done

**Review by nabinhait — APPROVED** [review]

**Pull request merged**